### PR TITLE
#92 Add few charts to show user stats

### PR DESCRIPTION
**Pull request by pramodbontha**
#92 changes
Added 3 blocks 
1.User Report - Showing general user score info.
2. Score Distribution chart - Shows score distribution of all users.
3. User progress chart - Shows progress on each quiz similar to time series.

Could you please let me know if any more charts needed to be added to the user.
Thank you in advance.

userstatistics

### CODE REVIEW

File: frontend/src/translations/de/common.json [hunk 1]
```diff
@@ -45,10 +47,17 @@
...
     "Submit Quiz": "Quiz einreichen",
     "Total Score": "Gesamtpunktzahl",
+    "Total Quizzes": "Total Quizze",
     "User": "Nutzer",
     "username": "Login-Name",
+    "User Report": "Benutzer-Bericht",
+    "User Progress by Every Quiz": "Benutzerfortschritt bei jedem Quiz",
+    "User Statistics": "Benutzer-Statistik",
+    "Score in percentage": "Punktzahl in Prozent",
+    "Score Distribution - All Users": "Partiturverteilung - Alle Benutzer",
```
Comment (tillmo) [inline frontend/src/translations/de/common.json:57]: "Partiturverteilung" should be "Punkteverteilung"

**Comment by ibieber** [comment]
To show how many students solve quizzes per day, I need:
- Two groups: Tester (admin, till, ibieber) and real students
- a bar chart with the number of students who have completed a quiz per day (since 20.04.2020) - shown per month

**Comment by pramodbontha** [comment]
> sorry, instead of candles, I meant box plots, see https://en.wikipedia.org/wiki/Box_plot
> The x axis of the score histogram is still not sorted by score

Professor, I could not find any box plot chart libraries free versions. I searched for the yesterday, all of them needs to be paid to use those libraries.
Shall i use another kind of chart. Could you please let me know if its fine

Thank you in advance

**Comment by pramodbontha** [comment]
> To show how many students solve quizzes per day, I need:
> 
> * Two groups: Tester (admin, till, ibieber) and real students
Hello @ibieber , Could you please give more inputs on this task. A simple mock up would help me understand clear on the representation of 2 groups in a bar chart.

Thank you in advance
> * a bar chart with the number of students who have completed a quiz per day (since 20.04.2020) - shown per month

**Comment by tillmo** [comment]
for drawing box plots see e.g. 
- https://github.com/yldio/chartjs-chart-box-plot based on https://en.wikipedia.org/wiki/Chart.js , see also https://www.chartjs.org/
- https://steemit.com/utopian-io/@faad/tutorial-10-dive-into-dc-js-a-javascript-library-box-plot based on https://github.com/dc-js/dc.js , see also https://dc-js.github.io/dc.js/

**Comment by tillmo** [comment]
In response to your conversation with @ibieber, I suggest to display a histogram for the average number of questions per week that a student has solved.

**Comment by tillmo** [comment]
looks good! I still have some comments, based on how the diagrams look with the real data. The first histogram looks OK now, but there could be a bit more space between the x axis inscriptions:
User Statistics charts (1)
I do not understand the x axis of the score graph (few values, and some are duplicates):
User Statistics charts (3)
The box plots are fine, but the x axis should be ordered according to quiz numbers, and you could omit the "quiz" prefix for brevity:
User-performance
The histogram at the bottom shows too few users (compare the first histogram):
User Statistics charts (4)

**Comment by pramodbontha** [comment]
I understand issues for the first 3 graphs.
@tillmo Could you please correct me if I am wrong.
For the 4 th graph, I am displaying the data as follow.
on the x-axis i am displaying quiz-id and for each 
quiz(i) = number of questions solved by all users for quiz(i)/all users attempted quiz(i)
So, i get average number of questions solved by users for that quiz and that is displayed on y-axis.
Could you please let me know if this is wrong.

Thank you in advance.

**Comment by tillmo** [comment]
> I understand issues for the first 3 graphs.
> @tillmo Could you please correct me if I am wrong.
> For the 4 th graph, I am displaying the data as follow.
> on the x-axis i am displaying quiz-id and for each
> quiz(i) = number of questions solved by all users for quiz(i)/all users attempted quiz(i)
> So, i get average number of questions solved by users for that quiz and that is displayed on y-axis.
> Could you please let me know if this is wrong.
> 
> Thank you in advance.

you are right. I have had a wrong idea about the 4th graph. Your formula is correct. Still, I wonder why for questions 5,6,11 etc. the result is zero here. When looking at the box plots, the result should be above zero.

**Comment by pramodbontha** [comment]
I was able to fix first 3 graph issues.
@tillmo Could you please suggest on the below
For the 4th graph issue. For the average question solved by a user, I am using only the correct answers count for every quiz. Maybe that is why for some quizzes the average question solved is 0. 
Should I use the number of questions attempted rather correct answer count. If I have to use only correct answer count I will allow float values in the graph.

Thank you in advance.

**Comment by tillmo** [comment]
thanks, looks better!
2nd graph: I still do not understand what this graph shows. Could you please explain?
3rd graph: please re-add the "quiz" prefix, because now the numbers bump into each other
4th graph: using only correct answers is OK, but then the inscription should be "Average number of questions correctly solved by a user per quiz"
Aha, and the 3rd and 4th graph use numbers to refer to quizzes. Could you please add those numbers also under "Available quizzes"? (Otherwise, it is unclear which quiz is meant by a certain number.)

**Comment by pramodbontha** [comment]
In 2nd graph, I tried to show how is user performing by every quiz. By every quiz, user will able to understand whether he performed better than previous quiz or not.

**Comment by tillmo** [comment]
but then why does the x axis read 3 3 1 2 1 1 in the above figure?

**Comment by tillmo** [comment]
Aha, the 2nd graph is about the logged-in user only. I suggest to have to separate statistics pages: one with the data of the logged-in user (i.e. 2nd graph and user report), an the other one with data concerning the entirety of users.

**Comment by tillmo** [comment]
the separation looks good! 
In the diagram "User Progress by Every Quiz", for quiz 1, I get 125%. How can this be?
The box plot should look like the image above in this pull request. Currently, the width seems to be too small, such that the inscriptions overlap.

**Comment by pramodbontha** [comment]
In the latest commit, I have added pagination similar to quiz questions where it shows up to only 10 quizzes in the box chart panel such that inscriptions do not overlap. Next and back buttons helps to navigate through next and previous 10 quizzes.

Regarding 125%, maybe because of the cache issue before, for one quiz there might have been multiple entries in the database.
While trying to calculate the score percentage I am summing up the score by grouping based on quiz id. For the earlier quizzes this might be a problem. For the new quizzes which are used after solving cache issue, they should be correct in percentages.

**Comment by tillmo** [comment]
looks good! Could you please increase the number of box plot items from 10 to 15 per page? Then it is fine.

**Review by tillmo — CHANGES_REQUESTED** [review]
for the box plots, on the first page I get quizzes 1-11, on the second page, quizzes 16-26. Please correct this

**Review by tillmo — APPROVED** [review]

**Pull request merged**